Add optimization docs (#68)
cossacklabs/product-docs: acra/configuring-maintaining/optimisations/acrastructs_vs_acrablocks.md

---
title: AcraStructs vs AcraBlocks
weight: 3
---

# AcraStructs vs AcraBlocks

There are two different "crypto containers" AcraServer can use to store encrypted value in.
Then, these containers are stored in database and decrypted back by AcraServer when needed.

History

> * `AcraStruct` is our "old" one, though we're not going to remove it any time soon.

Reviewer: It's not an old one. It is one of two. AcraStructs main crypto container for client-side encryption for now. Applications can encrypt with ephemeral asymmetric keys on app side with AcraStructs, but cannot do this with AcraBlocks because of symmetric key encryption. But they can configure AcraServer to re-encrypt AcraStructs (that generated on app side) on AcraServer side. Because AcraStructs allow to be sure that data are encrypted when it is transmitted between app and AcraServer.

Author: Rewrote this file a lot

* `AcraBlock` on the other hand is new and proposes some extra features.

Encryption with a key

> * `AcraStruct` uses key provided from somewhere outside (by Acra) to encrypt the value.

Reviewer: uses storage public key that related to private key known only for AcraServer. Main difference between AcraStruct and AcraBlock that first uses asymmetric key encryption that are slow due to multiplications of big numbers (elliptic curve encryption) and symmetric key encryption that just XOR arrays using optimized instructions in CPU.

* `AcraBlock` uses random key to encrypt the value, then encrypts this key with another key, this time provided from outside.
  That first key is called DEK (data encryption key) and the second one is called KEK (key encryption key).

Key rotation (when you need re-encrypt data with a new key)

> * With `AcraStruct`, the value has to be decrypted completely, then encrypted with a new key.

Reviewer: actually AcraStructs uses KEK and DEK too. But for KEK used asymmetric encryption that are slow. And you need to load private key, extract public key from AcraStruct, compute ECDHE to get "shared key" that used as KEK. And after that decrypt DEK.
When for AcraBlock AcraServer uses symmetric key and just decrypts KEK that encrypted and stored in AcraBlock. And decryption uses symmetric encryption that are much faster than asymmetric.

Author: Turned out I don't know their internal layout so well. Gonna update based on description from "Data structures" page.

* With `AcraBlock`, only DEK should be re-encrypted, leaving the encrypted value the same.
  And this may have a big impact on performance when rotating keys for a large amount of encrypted data.

Storage overhead

> * `AcraStruct` uses a bit less extra bytes compared to `AcraBlock`.

Reviewer: hm.... i'm not sure that AcraStructs uses less than AcraBlock.
```
package main

import (
	"fmt"
	"github.com/cossacklabs/acra/acrablock"
	"github.com/cossacklabs/acra/acrastruct"
	"github.com/cossacklabs/themis/gothemis/keys"
)
import _ "github.com/lib/pq"

func main() {
	testData := []byte(`a`)
	ecKeys, err := keys.New(keys.TypeEC)
	if err != nil {
		panic(err)
	}
	acraStruct, err := acrastruct.CreateAcrastruct(testData, ecKeys.Public, nil)
	if err != nil {
		panic(err)
	}
	acraBlock, err := acrablock.CreateAcraBlock(testData, []byte(`some key`), nil)
	if err != nil {
		panic(err)
	}
	fmt.Println(len(acraStruct), len(acraBlock))
}
// >> 190 139
```

AcraStruct: 190
AcraBlock 139

Author: Changed, is new version better?
(talking about storage overhead comparison, other issues in this file will be fixed in next commit)

---
title: AcraStructs vs AcraBlocks
weight: 3
---

# AcraStructs vs AcraBlocks

There are two different "crypto containers" AcraServer can use to store encrypted value in.
Then, these containers are stored in database and decrypted back by AcraServer when needed.

> `AcraStruct` is the main crypto container for the client-side encryption.

Reviewer 2: I'd leave `` `backticks` `` flair for code examples and other verbatim uses. I think using PascalCase is enough to make AcraStruct to stand out as a term.

On a side note, while looking for a way to escape those, here's what I found out about GitHub parser:

### <code><code><code><code><code><code><code><code><code><code><code><code><code><code><code><code>code</code></code></code></code></code></code></code></code></code></code></code></code></code></code></code></code>

By the way, the string above is

    `` `backticks` ``

> `AcraStruct` is the main crypto container for the client-side encryption.
> Applications can encrypt with ephemeral asymmetric keys on their side with `AcraStruct`s,
> but cannot do this with `AcraBlock`s because of symmetric key encryption.

Reviewer 2: This reads like blaming AcraBlocks for not doing what they were explicitly designed not to do.

```suggestion
AcraStruct is the main crypto container for the client-side encryption.
AcraStructs use ephemeral asymmetric keys to enable applications to encrypt the data but never decrypt it –
the private key stays on the AcraServer which is the only one capable of decryption.
```

And you can add a blurb about AcraBlocks in a separate paragraph after this one. For example,

```markdown
AcraBlocks use only symmetric cryptography to they are generally faster then AcraStructs.
However, this comes at a cost of having to do stricter key management,
since now the application has the decryption keys as well which is a risk.
```

You can also configure AcraServer to re-encrypt `AcraStruct`s (that were generated on app side)
on AcraServer side, this way data will definitely be encrypted while transmitted between

> some client app and AcraServer.

Reviewer 2: > this way data will definitely be encrypted while transmitted between some client app and AcraServer.

It's it already encrypted by the transport layer in flight in case of transparent encryption?

I think the point of reencrypting AcraStructs is to make sure that you can't match anything known to the application to the data stored in the database. So an attacker can't submit some AcraStruct, expect it to be stored verbatim, then peek into the database for what row it has been stored at, and then learn something about the system from that.

Another point is preventing usage of compromised keys. AcraServer might accept AcraStructs encrypted with old keys (not necessarily compromised but rotated out), and then reencrypt the data with the current ones.

@Lagovas, @storojs72, what to you think about this part of the paragraph?

Reviewer 3: > You can also configure AcraServer to re-encrypt AcraStructs (that were generated on app side) on AcraServer side, this way data will definitely be encrypted while transmitted between some client app and AcraServer

I agree, it sounds a bit odd. If AcraStruct came to AcraServer from client app, then plaintext from app is definitely encrypted - so there is no need to additionally ensure that data is encrypted. @ilammy's explanation sounds more relevant - with mentioned re-encryption attacker is not able to push specifically crafted AcraStructs to AcraServer and then observe them in the database, since after re-encryption AcraStruct will have different value.

Reviewer: +1, you are right

### Encryption with a key

> 

Reviewer 2: I don't really like this header (it's don't say anything to me), but I don't have any real suggestions either. IDK, “Key wrapping”? Something that *actually* relates to the matter discussed in this section. “Encryption with a key” is not it. It's obvious that you need a key for encryption. But the section is not about that, it's about how both containers have a DEK but they deal with KEK differently.

Both containers generate random symmetric data encryption key (DEK) and encrypt the data with it.
This key is then encrypted with another key, and here is the difference:
* `AcraStruct` uses asymmetric cryptography to encrypt DEK, one ephemeral keypair and public key
  for which Acra knows the private key, this private key will then be used to decrypt the data;
  this method is also a bit slower because of the usage of asymmetric cryptography
* `AcraBlock` uses one more symmetric layer to encrypt the DEK, the same key will be used for
  both encryption and decryption

### Key rotation (when you need re-encrypt data with a new key)

Both crypto containers have ability to re-encrypt DEK without touching the encrypted data.
However, `AcraStruct` will generally be slower because it has to perform more computation
to decrypt the DEK and encrypt it again with another ephemeral keypair.
As was mentioned above, this is because of asymmetric cryptography being used (need to compute
shared ECDH secret, generate ECDH keypair etc).

> ### Storage overhead (not counting encrypted data length)
> * `AcraStruct` takes about `189` additional bytes
> * `AcraBlock` takes about `138` additional bytes

Reviewer 2: “Takes” usually means in “in total” but here we're talking about overhead, so how about this?

```suggestion
### Storage overhead (in addition to encrypted data length)

* `AcraStruct` adds about `189` extra bytes
* `AcraBlock` adds about `138` extra bytes
```
